### PR TITLE
Add animation parameters to allow customization of Group updates.

### PR DESCRIPTION
**Pull request by lepond**
## Change summary
- Update `Group` `_setAnimations` to take `SpringAnimationParameters` to allow more control over how constrainables are animated in or out.

## How was it tested?
*How did you verify that this change accomplished what you expected? Add more detail as needed.*
- [ ] Wrote automated tests
- [X] Built and ran on the iOS simulator
- [ ] Built and ran on a device

## Pull request checklist
*All items in this checklist must be completed before a pull request will be reviewed.*

- [ ] Risky changes have been put behind a feature flag, e.g. `CollectionViewConfiguration`
- [x] Added a [`CHANGELOG.md` entry](https://keepachangelog.com/en/1.0.0/) in the "Unreleased" section for any library changes

### CODE REVIEW

File: Sources/EpoxyLayoutGroups/Groups/VGroup.swift
Comment (thedrick) [inline Sources/EpoxyLayoutGroups/Groups/VGroup.swift:110]: can we include the same comment here that we have in `HGroup`?

**Review by thedrick — APPROVED** [review]
This looks great @lepond thank you so much!

File: Sources/EpoxyLayoutGroups/Groups/VGroup.swift
Comment (lepond) [reply Sources/EpoxyLayoutGroups/Groups/VGroup.swift:117]: Sure thing. Done!

**Comment by lepond** [comment]
@thedrick - would you be able to help me kick off the running workflows? 
I don't have permission because this is my first contribution.

File: Sources/EpoxyLayoutGroups/Types/SpringAnimationParameters.swift
Comment (erichoracek) [inline Sources/EpoxyLayoutGroups/Types/SpringAnimationParameters.swift:7]: Now that we're making this a public API, I think I would prefer that we write this in a slightly different way, e.g.:
```swift
public struct LayoutGroupUpdateAnimation: Hashable {
  public init(…) { … }
  public var id: AnyHashable
  public var animate: (_ animations: () -> Void, _ completion: (_ completed: Bool) -> Void)
}
```
Then we could define spring animations something like:
```swift
extension LayoutGroupUpdateAnimation {
  public static func spring(
    duration: TimeInterval = 0.5, 
    delay: TimeInterval = 0.0,
    dampingRatio: CGFloat = 1.0, 
    initialSpringVelocity: CGFloat = 0.0)
  {
    .init(id: AnimationID.spring) { animations, completion in
      UIView.animate(
          withDuration: duration,
          delay: delay,
          usingSpringWithDamping: dampingRatio,
          initialSpringVelocity: initialSpringVelocity,
          options: [.beginFromCurrentState, .allowUserInteraction],
          animations: animations
          completion: completion)
    }
  }
}

private enum AnimationID {
  case spring
}
```
Which enables full flexibility, e.g. if someone wants a non-spring animation

**Review by erichoracek — COMMENTED** [review]
@lepond Have a question/suggestion on the overall API shape cc @thedrick

File: Sources/EpoxyLayoutGroups/Types/SpringAnimationParameters.swift
Comment (lepond) [reply Sources/EpoxyLayoutGroups/Types/SpringAnimationParameters.swift:7]: @erichoracek -- recommendations on hashability? Is there a way to ignore the hashability on the closure? Or to ignore the LayoutGroupUpdateAnimation on the style?

Comment (erichoracek) [reply Sources/EpoxyLayoutGroups/Types/SpringAnimationParameters.swift:7]: @lepond That's what I was imagining that the `id` would serve as, so the `Hashable` implementation would look something like:
```swift
extension LayoutGroupUpdateAnimation {
  public func hash(into hasher: inout Hasher) {
    hasher.combine(id)
  }

  public static func == (lhs: LayoutGroupUpdateAnimation, rhs: LayoutGroupUpdateAnimation) -> Bool {
    lhs.id == rhs.id
  }
}
```

File: CHANGELOG.md
Comment (erichoracek) [inline CHANGELOG.md:27]:
```suggestion
- Added `LayoutGroupUpdateAnimation` for customizing `Group` animated updates.
```

File: Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift
Comment (erichoracek) [inline Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift:61]: Can we make this type private? This should also allow us to also remove the explicit initializer

File: Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift
Comment (erichoracek) [inline Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift:59]: Apologies if my comment was unclear/edit, I was thinking to use `SpringAnimationParameters` as the ID since it includes all of the fields in its equality, rather than `AnimationID.spring`, so I think we can remove this type entirely

File: Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift
Comment (erichoracek) [inline Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift:49]: nit: I think we can omit this return
```suggestion
    lhs.id == rhs.id
```

File: Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift
Comment (erichoracek) [inline Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift:24]: nit: I'm not quite sure what static means here, can we omit?
```suggestion
  ///   - duration: a TimeInterval for animation duration, with a default value of `0.5`
  ///   - delay: a TimeInterval for animation duration, with a default value of `0.0`
  ///   - dampingRatio: a value for spring damping ratio, with a default value of `1.0`
  ///   - initialSpringVelocity: a value for initial spring velocity, wiht a default value of `0.0`
```

File: Sources/EpoxyLayoutGroups/Types/GroupAnimationStyle.swift
Comment (erichoracek) [inline Sources/EpoxyLayoutGroups/Types/GroupAnimationStyle.swift:10]: We already have a way to specify no animation—specifically you pass `animated: false` to the `setItems(…)` method, so I'm not sure we need this `GroupAnimationStyle` type and we can just have the `LayoutGroupUpdateAnimation`

File: Sources/EpoxyLayoutGroups/Groups/Group.swift
Comment (erichoracek) [inline Sources/EpoxyLayoutGroups/Groups/Group.swift:105]: I think we still want the `animated: Bool` flag passed to this method, and we would just pass an additional `animationStyle` that would decide whether it was animated or not that comes from the `Group.animation`

**Review by erichoracek — COMMENTED** [review]
@lepond Thanks for making the updates, have some suggestions/questions on the changes

File: Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift
Comment (lepond) [reply Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift:61]: All the values in this spring animation are what I need to customize on the style of the appear animation I am adding...

File: Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift
Comment (lepond) [reply Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift:59]: Okay - that makes more sense.

File: Sources/EpoxyLayoutGroups/Groups/Group.swift
Comment (erichoracek) [inline Sources/EpoxyLayoutGroups/Groups/Group.swift:105]: I believe we can make the `animation` parameter required right?

File: Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift
Comment (erichoracek) [inline Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift:6]: I think we can omit the "closure" suffix on this since it's duplicative with the type info and then add some documentation comments
```suggestion
/// A closure that's invoked to perform an animated update to the items in a layout group,
/// with an animations closure containing the updates to animate and a completion closure
/// to be invoked when the animations complete with a parameter of whether the animations
/// completed successfully.
public typealias GroupAnimation = (
  _ animations: @escaping () -> Void, 
  _ completion: @escaping (_ completed: Bool) -> Void) 
  -> Void
```

File: Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift
Comment (erichoracek) [inline Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift:8]: Since the `_setItems` method is private let's only document the public interface
```suggestion
/// Provides parameter values, which can be passed to the initializer of a Group
```

File: Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift
Comment (erichoracek) [inline Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift:15]:
```suggestion
  /// An identifier that uniquely identifies this update animation.
  public var id: AnyHashable
```

File: Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift
Comment (erichoracek) [inline Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift:16]:
```suggestion
  /// A closure that's called to perform the update animations.
  public var animate: GroupAnimationClosure
```

**Review by erichoracek — APPROVED** [review]
Looking good, almost there! Just a few small comments

File: Sources/EpoxyLayoutGroups/Groups/HGroup.swift
Comment (erichoracek) [inline Sources/EpoxyLayoutGroups/Groups/HGroup.swift:137]: Looks like we're missing a documentation comment on this new parameter above

File: Sources/EpoxyLayoutGroups/Groups/Group.swift
Comment (lepond) [reply Sources/EpoxyLayoutGroups/Groups/Group.swift:105]: Yes - and probably best to make it required so that there is an actual animation to render when animated is `true`!

**Comment by lepond** [comment]
hey @erichoracek - what are the next steps for merging? (I do not have permissions because I am not yet a collaborator.)

File: Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift
Comment (erichoracek) [inline Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift:17]:
```suggestion
  public init(id: AnyHashable, animate: @escaping GroupAnimation) {
```

File: Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift
Comment (erichoracek) [inline Sources/EpoxyLayoutGroups/Types/LayoutGroupUpdateAnimation.swift:25]:
```suggestion
  public var animate: GroupAnimation
```

**Review by erichoracek — APPROVED** [review]
@lepond this isn't building currently but with a few changes we should be good to go! Once the CI checks are passing we can merge

**Pull request merged**
